### PR TITLE
DPDV-4549: Do not crash when invalid query is used

### PR DESCRIPTION
**Pull request by martin-majlis-s1**
Jira Link: <https://sentinelone.atlassian.net/browse/DPDV-4549>
Jira Link: <https://sentinelone.atlassian.net/browse/DPDV-4411>

--------

# 🥅 Goal

In the PR https://github.com/scalyr/dataset-addon-for-splunk/pull/15/files#diff-ad6d3503f1efe6044e12096a99eaa00b173327c48ca60dbaebea7630451e8e9eL220 we have switched to LRQ API and in this transition we have lost also error reporting. So lets introduce error reporting back.

# 🛠️ Solution

## Propagate error messages

Old API was returning errors as user friendly messages - `{'message': 'Error in filter expression: value expected', 'status': 'error/client/badParam'}` - and therefore UI was just showing the `message` field. The new API is returning less user friendly error messages - `{"code":"invalid_argument","message":"field=[filter] error=[value expected]","details":[{"field":"filter","message":"value expected"}]}`

## Change default to 15 minutes

I am testing this against QA environment. It has some problems, that it almost always fails when I try to search for 4 hours of data. The consequence is that the `serverHost` field is never populated, and therefore `Base Query` remains empty.
Screenshot 2023-11-01 at 12 43 09
Screenshot 2023-11-01 at 12 45 34
Which is invalid query => all the examples fails.
Screenshot 2023-11-01 at 12 47 01

It's really bad experience, when you go to the example page, and you see only errors, because it's not able to fetch 4 hours of data. So lets be on the safe side.

# 🏫 Testing

## Search

Screenshot 2023-10-31 at 15 39 48

## 500 in the middle of long running query

https://github.com/scalyr/dataset-addon-for-splunk/pull/96#issuecomment-1788781557
Screenshot 2023-11-01 at 12 14 30

## Timeseries

Time series is still using the old API, but the error reporting part was also missing. Query with [serverHost=](http://localhost:8000/en-GB/app/TA_dataset/search?q=%7C%20dataset%20method%3Dtimeseries%20search%3D%22serverHost%3D%22%20function%3D%22rate%22%20buckets%3D24%20createsummaries%3Dfalse%20onlyusesummaries%3Dfalse%20%7C%20spath%20%7C%20timechart%20values(rate)%20as%20rate&display.page.search.mode=smart&dispatch.sample_ratio=1&workload_pool=&earliest=-15m&latest=now&display.page.search.tab=statistics&display.general.type=statistics&sid=1698831175.101)
Screenshot 2023-11-01 at 10 33 41

### CODE REVIEW

File: TA_dataset/bin/dataset_api.py [hunk 1]
```diff
@@ -82,34 +103,42 @@ def ds_lrq_facet_values(
...
+            )
+            result = response.parsed
+            steps_done = result.steps_completed
+        delete_queries.sync_detailed(
+            id=query_id, client=client, forward_tag=forward_tag
+        )
 
-    return result
+        return result, None
+    return None, json.loads(response.content)
```
Comment (tomaz-s1) [inline TA_dataset/bin/dataset_api.py:141]: So we return the whole JSON response body as an error? Do we want to return just a specific field or similar?

**Review by tmartin-s1 — COMMENTED** [review]
I realize it's not perfect, but at least presenting this as in the screenshot is better as it will give the user a better clue as to what the actual problem is.

File: TA_dataset/bin/dataset_api.py (hunk 1 above)
Comment (martin-majlis-s1) [reply TA_dataset/bin/dataset_api.py:141]: The error is passed to the function - `search_error_exit` - https://github.com/scalyr/dataset-addon-for-splunk/blob/main/TA_dataset/bin/dataset_search_command.py#L119 - which expects dictionary, where there is attribute `message` - and this is the same in both cases.

**Comment by martin-majlis-s1** [comment]
The new API is behaving little bit strange. When I ask for 1 week of data, it very often crashes.

```
2023-11-01 11:02:37,174 DEBUG pid=73847 tid=MainThread file=dataset_api.py:ds_lrq_run_loop:130 | Response(status_code=<HTTPStatus.OK: 200>, content=b'{"id":"eyJ0eXBlIjoiUFEiLCJ0b2tlbiI6ImQyY2M5MGYzLTc4MTEtNDYwYy04ZDJjLTVjMTk3N2I5NWU5ZCJ9","stepsCompleted":14,"stepsTotal":16,"resolvedTimeRange":{"start":1698231600000000000,"end":1698836473000000000},"data":null,"totalSteps":16}', headers=Headers({'server': 'nginx', 'date': 'Wed, 01 Nov 2023 11:02:37 GMT', 'content-type': 'application/json;charset=UTF-8', 'content-length': '229', 'connection': 'keep-alive', 'set-cookie': 'sp=cb95e6d4-ba84-4c23-ad47-c2309b36e19d;Expires=Thu, 31-Oct-2024 16:51:22 GMT;Max-age=31556926;path=/;HttpOnly', 'x-dataset-query-forward-tag': 'TAG-', 'expires': 'Thu, Jan 1 2009 12:00:00 GMT', 'cache-control': 'no-cache, must-revalidate', 'pragma': 'no-cache', 'scalyr-team-token': 'TOKEN--', 'access-control-allow-credentials': 'true', 'access-control-allow-methods': 'GET, POST, PUT, DELETE, OPTIONS', 'access-control-allow-headers': 'Accept,Authorization,Cache-Control,Content-Type,DNT,If-Modified-Since,Keep-Alive,Origin,User-Agent,X-Requested-With', 'access-control-max-age': '1728000'}), parsed=QueryResult(id='eyJ0eXBlIjoiUFEiLCJ0b2tlbiI6ImQyY2M5MGYzLTc4MTEtNDYwYy04ZDJjLTVjMTk3N2I5NWU5ZCJ9', steps_completed=14, steps_total=16, resolved_time_range=TimeRangeResultData(start=1698231600000000000, end=1698836473000000000, additional_properties={}), error=<dataset_query_api_client.types.Unset object at 0x7ffffd5f8190>, data=None, additional_properties={'stepsTotal': 16}))
2023-11-01 11:02:37,655 DEBUG pid=73847 tid=MainThread file=dataset_api.py:ds_lrq_run_loop:130 | Response(status_code=<HTTPStatus.INTERNAL_SERVER_ERROR: 500>, content=b'{"code":"internal_server_error","message":"Operation not permitted. Operation not permitted. You do not have access to this account.","details":[]}', headers=Headers({'server': 'nginx', 'date': 'Wed, 01 Nov 2023 11:02:37 GMT', 'content-type': 'application/json;charset=UTF-8', 'content-length': '147', 'connection': 'keep-alive', 'set-cookie': 'sp=3240c1be-f9cf-4d22-b2ff-f8dfdb195354;Expires=Thu, 31-Oct-2024 16:51:23 GMT;Max-age=31556926;path=/;HttpOnly', 'x-dataset-query-forward-tag': 'TAG-', 'expires': 'Thu, Jan 1 2009 12:00:00 GMT', 'cache-control': 'no-cache, must-revalidate', 'pragma': 'no-cache', 'scalyr-team-token': 'TOKEN--', 'access-control-allow-credentials': 'true', 'access-control-allow-methods': 'GET, POST, PUT, DELETE, OPTIONS', 'access-control-allow-headers': 'Accept,Authorization,Cache-Control,Content-Type,DNT,If-Modified-Since,Keep-Alive,Origin,User-Agent,X-Requested-With'}), parsed=None)
```

I can see, that it was able to fetch first 14 chunks and then it fails on Internal Server Error 500 - `Operation not permitted. Operation not permitted. You do not have access to this account.`

File: TA_dataset/bin/dataset_api.py [hunk 2]
```diff
@@ -25,6 +24,11 @@
 )
 
 
+class APIException(Exception):
+    def __init__(self, payload):
+        self.payload = payload
```
Comment (tomaz-s1) [inline TA_dataset/bin/dataset_api.py:29]: Would it also make sense to store other information such as the response status code?

File: TA_dataset/bin/dataset_api.py [hunk 3]
```diff
@@ -131,14 +127,27 @@ def ds_lrq_run_loop(
...
                 forward_tag=forward_tag,
             )
+            logger().debug(response)
             result = response.parsed
-            steps_done = result.steps_completed
+            if result:
+                steps_done = result.steps_completed
+                retry = 0
+            else:
+                retry += 1
```
Comment (tomaz-s1) [inline TA_dataset/bin/dataset_api.py:136]: Eventually it would be nice to abstract retries into the client library itself :)

File: TA_dataset/bin/dataset_api.py (hunk 2 above)
Comment (tomaz-s1) [reply TA_dataset/bin/dataset_api.py:29]: Please also add a comment on what `payload` contains / stores - API response body.

File: TA_dataset/bin/dataset_search_command.py [hunk 4]
```diff
@@ -116,17 +118,19 @@ def get_search_arguments(self):
     )
 
 
-def search_error_exit(self, r_json):
+def search_error_exit(self, r_json: Union[Dict[str, Any], str]) -> None:
```
Comment (tomaz-s1) [inline TA_dataset/bin/dataset_search_command.py:121]: Do we have any unit tests for this function?

File: TA_dataset/bin/dataset_api.py (hunk 2 above)
Comment (martin-majlis-s1) [reply TA_dataset/bin/dataset_api.py:29]: Currently it's passed to `search_error_exit` which does not care. I am not sure, whether it's useful for user to know, that it was 401 or 503. I hope that this is captured in that message.

File: TA_dataset/bin/dataset_api.py (hunk 3 above)
Comment (martin-majlis-s1) [reply TA_dataset/bin/dataset_api.py:136]: My observation is, that when it gets to the 500 status, it's not resolved by these subsequent calls. So it could be most likely removed.

I will add comment.

File: TA_dataset/bin/dataset_search_command.py (hunk 4 above)
Comment (martin-majlis-s1) [reply TA_dataset/bin/dataset_search_command.py:121]: I have added the unit test.

File: TA_dataset/bin/dataset_api.py [hunk 5]
```diff
@@ -87,29 +107,53 @@ def ds_lrq_facet_values(
...
+        retry = 0
+        while steps_done < steps_total:
+            response = get_queries.sync_detailed(
+                id=query_id,
+                query_type=body.query_type,
+                client=client,
+                last_step_seen=steps_done,
+                forward_tag=forward_tag,
+            )
+            logger().debug(response)
```
Comment (tomaz-s1) [inline TA_dataset/bin/dataset_api.py:132]: Should we also include some message in addition to the response object? Or does response object str representation already include all / enough context?

File: TA_dataset/bin/dataset_api.py [hunk 6]
```diff
@@ -87,29 +107,53 @@ def ds_lrq_facet_values(
...
-        )
-        result = response.parsed
+    if result:
+        forward_tag = response.headers["x-dataset-query-forward-tag"]
         steps_done = result.steps_completed
-    delete_queries.sync_detailed(id=query_id, client=client, forward_tag=forward_tag)
+        steps_total = result.steps_total
+        query_id = result.id
+        retry = 0
+        while steps_done < steps_total:
```
Comment (tomaz-s1) [inline TA_dataset/bin/dataset_api.py:124]: Some tests for retry functionality (including edge cases would be good), besides that, LGTM

**Review by tomaz-s1 — APPROVED** [review]
LGTM, but let's please add some test cases for retry functionality where we get a chance.

File: TA_dataset/bin/dataset_api.py (hunk 5 above)
Comment (martin-majlis-s1) [reply TA_dataset/bin/dataset_api.py:132]: I do not know, whether we are receiving logs from somebody. So it's for my development, so I do not need more context.

File: TA_dataset/bin/dataset_api.py (hunk 6 above)
Comment (martin-majlis-s1) [reply TA_dataset/bin/dataset_api.py:124]: There is no complicated logic with exponential backoff or similar. It's very naive implementation. By eyeballing it one more time, I am convinced, that it cannot end up in the infinite loop.
